File: lib/evidenceIndex/enrichers/stormOverlay.js
// lib/evidenceIndex/enrichers/stormOverlay.js
//
// Real integration with the National Weather Service API
// (https://api.weather.gov) — free, keyless, matches the data source
// AeroLeadAI's own weather-agent already uses. Pulls active alerts for a
// lat/lon, extracts hail size / wind speed where the alert text actually
// states them (real regex against real NWS alert text, not invented), and
// flags heavy snow/rain from the alert's event type. Ported from
// evidence_index/ Python prototype's enrichers/storm_overlay.py, unchanged.
//
// NWS active alerts only cover CURRENT/recent weather, not deep historical
// hail swaths — for that you'd need NOAA's Storm Events Database (a bulk
// CSV/API dataset, different product, not wired here).

import { NWS_API_BASE, NWS_USER_AGENT } from "../config";

const HEADERS = { "User-Agent": NWS_USER_AGENT, Accept: "application/geo+json" };

const HAIL_SIZE_PATTERN = /(\d+(?:\.\d+)?)\s*(?:inch|in)\b.*?hail/i;
const WIND_SPEED_PATTERN = /(\d+)\s*mph/i;

/**
 * Real call to NWS active alerts for a point. Returns [] on any failure
 * rather than throwing — a storm overlay that can't reach the API should
 * degrade to "no storm evidence found", not crash the pipeline.
 */
export async function getActiveAlerts(lat, lon) {
  try {
    const params = new URLSearchParams({ point: `${lat},${lon}` });
    const res = await fetch(`${NWS_API_BASE}/alerts/active?${params.toString()}`, {
      headers: HEADERS,
      signal: AbortSignal.timeout(10000),
    });
    if (!res.ok) throw new Error(`NWS API returned ${res.status}`);
    const json = await res.json();
    return json?.features || [];
  } catch (e) {
    console.warn(`[stormOverlay] NWS API call failed: ${e.message}`);
    return [];
  }
}

/**
 * Extracts hailInches / windMph / heavySnow / heavyRain from real alert
 * text. Only sets a value when the text actually supports it — never
 * defaults to a nonzero guess.
 */
export function parseStormSignals(alerts) {
  let hailInches = 0;
  let windMph = 0;
  let heavySnow = false;
  let heavyRain = false;

  for (const alert of alerts) {
    const props = alert?.properties || {};
    const event = (props.event || "").toLowerCase();
    const text = `${props.description || ""} ${props.headline || ""}`;

    const hailMatch = HAIL_SIZE_PATTERN.exec(text);
    if (hailMatch) hailInches = Math.max(hailInches, parseFloat(hailMatch[1]));

    const windMatch = WIND_SPEED_PATTERN.exec(text);
    if (windMatch) windMph = Math.max(windMph, parseInt(windMatch[1], 10));

    if (event.includes("snow") || event.includes("blizzard")) heavySnow = true;
    if (event.includes("flood") || event.includes("rain")) heavyRain = true;
  }

  return {
    hailInches,
    windMph,
    heavySnow,
    heavyRain,
    stormConfidence: alerts.length ? 90 : 0, // real alerts = high confidence in the storm signal; no alerts = no signal, not "no storm"
  };
}

/**
 * Expects lead.lat / lead.lon. If either is missing, returns the lead
 * unchanged (plus stormConfidence: 0) rather than guessing a location.
 */
export async function enrichStorm(lead) {
  const { lat, lon } = lead;
  if (lat == null || lon == null) return { ...lead, stormConfidence: 0 };

  const alerts = await getActiveAlerts(lat, lon);
  const signals = parseStormSignals(alerts);

  // Only fill in storm fields the lead doesn't already have from another
  // source (e.g. a more precise hail-swath dataset) — never overwrite
  // better data with a coarser real-time alert parse.
  const merged = { ...lead };
  for (const [key, value] of Object.entries(signals)) {
    if (merged[key] === undefined || merged[key] === null || merged[key] === 0 || merged[key] === false) {
      merged[key] = value;
    }
  }

  return merged;
}

File: evidence_index/enrichers/storm_overlay.py
"""
enrichers/storm_overlay.py

Real integration with the National Weather Service API
(https://api.weather.gov) — free, keyless, matches the data source
AeroLeadAI's own weather-agent already uses. Pulls active alerts for a
lat/lon, extracts hail size / wind speed where the alert text actually
states them (real regex against real NWS alert text, not invented), and
flags heavy snow/rain from the alert's event type.

NWS active alerts only cover CURRENT/recent weather, not deep historical
hail swaths — for that you'd need NOAA's Storm Events Database (a bulk
CSV/API dataset, different product, not wired here). Flagged rather than
silently treated as equivalent.
"""

import re
import requests

from config import NWS_API_BASE, NWS_USER_AGENT

HEADERS = {"User-Agent": NWS_USER_AGENT, "Accept": "application/geo+json"}

HAIL_SIZE_PATTERN = re.compile(r"(\d+(?:\.\d+)?)\s*(?:inch|in)\b.*?hail", re.IGNORECASE)
WIND_SPEED_PATTERN = re.compile(r"(\d+)\s*mph", re.IGNORECASE)


def get_active_alerts(lat: float, lon: float) -> list[dict]:
    """Real call to NWS active alerts for a point. Returns [] on any
    failure rather than raising — a storm overlay that can't reach the
    API should degrade to 'no storm evidence found', not crash the pipeline."""
    try:
        resp = requests.get(
            f"{NWS_API_BASE}/alerts/active",
            params={"point": f"{lat},{lon}"},
            headers=HEADERS,
            timeout=10,
        )
        resp.raise_for_status()
        return resp.json().get("features", [])
    except requests.RequestException as e:
        print(f"[storm_overlay] NWS API call failed: {e}")
        return []


def parse_storm_signals(alerts: list[dict]) -> dict:
    """Extracts hail_inches / wind_mph / heavy_snow / heavy_rain from real
    alert text. Only sets a value when the text actually supports it —
    never defaults to a nonzero guess."""
    hail_inches = 0.0
    wind_mph = 0
    heavy_snow = False
    heavy_rain = False

    for alert in alerts:
        props = alert.get("properties", {})
        event = (props.get("event") or "").lower()
        description = props.get("description") or ""
        headline = props.get("headline") or ""
        text = f"{description} {headline}"

        hail_match = HAIL_SIZE_PATTERN.search(text)
        if hail_match:
            hail_inches = max(hail_inches, float(hail_match.group(1)))

        wind_match = WIND_SPEED_PATTERN.search(text)
        if wind_match:
            wind_mph = max(wind_mph, int(wind_match.group(1)))

        if "snow" in event or "blizzard" in event:
            heavy_snow = True
        if "flood" in event or "rain" in event:
            heavy_rain = True

    return {
        "hail_inches": hail_inches,
        "wind_mph": wind_mph,
        "heavy_snow": heavy_snow,
        "heavy_rain": heavy_rain,
        "storm_confidence": 90 if alerts else 0,  # real alerts = high confidence in the storm signal; no alerts = no signal, not "no storm"
    }


def enrich_storm(lead: dict) -> dict:
    """
    Expects lead to have 'lat' and 'lon'. If either is missing, returns
    the lead unchanged rather than guessing a location.
    """
    lat, lon = lead.get("lat"), lead.get("lon")
    if lat is None or lon is None:
        return {**lead, "storm_confidence": 0}

    alerts = get_active_alerts(lat, lon)
    signals = parse_storm_signals(alerts)

    # Only fill in storm fields the lead doesn't already have from another
    # source (e.g. a more precise hail-swath dataset) — never overwrite
    # better data with a coarser real-time alert parse.
    merged = {**lead}
    for key, value in signals.items():
        if key not in merged or merged[key] in (None, 0, False):
            merged[key] = value

    return merged
